s3_start_transfer/src/verify_transfer_packages.py

```diff
@@ -0,0 +1,321 @@
+# -*- encoding: utf-8
+"""
+The Archivematica transfer format is a bit fiddly, and there are ways to get
+it wrong that are non-obvious.
+
+This runs some checks over the package before sending it to Archivematica.
+
+It needs two things from the package:
+*   The list of all files in the package (from zipfile.namelist())
+*   The contents of ``metadata/metadata.csv`` (if present in the package)
+
+"""
+
+import csv
+import inspect
+import io
+import os
+import textwrap
+
+
+class VerificationFailure(Exception):
+    def __init__(self, message):
+        super().__init__(textwrap.dedent(message).strip())
+
+
+def verify_package(*, logger, zip_file):
+    # Extract the zip file listing and the metadata.csv contents for this
+    # transfer package.
+    file_listing = zip_file.namelist()
+
+    try:
+        metadata_csv = zip_file.open("metadata/metadata.csv")
+    except KeyError:
+        metadata = None
+    else:
+        metadata = metadata_csv.read().decode("utf8")
+
+        # Replace any byte-order marks in the CSV, we don't need them.
+        # These are sometimes written by Excel and the like, I think?
+        if "\ufeff" in metadata:
+            metadata = metadata.replace("\ufeff", "")
+
+    verifications = [
+        verify_all_files_not_under_single_dir,
+        verify_all_files_not_under_objects_dir,
+        verify_has_a_metadata_csv,
+        verify_only_metadata_csv_in_metadata_dir,
+        verify_metadata_csv_is_correct_format,
+    ]
+
+    logger.write(
+        f"Running {len(verifications)} checks for {zip_file}"
+    )
+
+    for i, verify_function in enumerate(verifications, start=1):
+        logger.write(f"== Check {i}: {verify_function.__name__} ==")
+
+        # This slightly hacky code is to get around the fact that not all
+        # functions need both the file listing and the metadata, and rather
+        # than have them take unused arguments, we build their signature
+        # dynamically.  Dynamic programming, eh?
+        kwargs = {}
+
+        if "file_listing" in inspect.getfullargspec(verify_function).args:
+            kwargs["file_listing"] = file_listing
+
+        if "metadata" in inspect.getfullargspec(verify_function).args:
+            kwargs["metadata"] = metadata
+
+        try:
+            verify_function(**kwargs)
+        except VerificationFailure as err:
+            logger.write("Check failed:\n")
+            logger.write(str(err) + "\n")
+            return False
+
+    logger.write("All checks complete!")
+
+    return True
+
+
+def verify_all_files_not_under_single_dir(file_listing):
```

Comment:
After reading this function name and description I understood it more clearly as `verify_all_files_at_top_level` rather than what they're not. Assuming this is named after a common cause of error though.